### PR TITLE
Provider rewrite

### PR DESCRIPTION
**Pull request by elijaharita**
The provider implementation I wrote originally did not work very well. Due to the structure of the old provider code (everything stored on goroutine stacks), it was very difficult to keep track of the connections between requests, retrievals, and responses, and many requests would never receive any response. I could not identify an easy solution working off of the existing code so I started from a blank slate. It works great now.

The new architecture keeps request, retrieval, and response tasks in three clean peer task queues. This makes it a lot easier for the provider to track what's going on, and requests are no longer dropped.

## TODO
A failed retrieval currently does not result in DONT_HAVE being sent. This is not a regression, because Autoretrieve has never done that. It will be a lot easier to add this now but I don't think it needs to be a part of this PR.

### Removed
- `max-bitswap-workers=[uint]`

### Added
- `request-workers=[uint]`
- `response-workers=[uint]`
- `retrieval-workers=[uint]`

### CODE REVIEW

**Comment by kylehuntsman** [comment]
Thanks for keeping the BitswapResponse metrics 👍

File: bitswap/provider.go [hunk 1]
```diff
@@ -94,264 +120,305 @@ func NewProvider(
...
+		return nil
+	}
+
+	stats.Record(ctx, metrics.BitswapRequestCount.M(1))
+
+	// First check blockstore, immediately write to response queue if it
+	// exists
+
+	size, err := provider.blockManager.GetSize(ctx, entry.Cid)
+	if err == nil {
```
Comment (rvagg) [inline bitswap/provider.go:217]: may be worth bringing back the not-not-found error logging for this as an `} else !ipld.IsNotFound(err) { ... logger.Warnf... }`

File: bitswap/provider.go [hunk 2]
```diff
@@ -94,264 +120,305 @@ func NewProvider(
...
 		switch entry.WantType {
 		case wantTypeHave:
-			// For WANT_HAVE, just confirm whether it's in the blockstore
-			has, err := provider.blockManager.Has(ctx, entry.Cid)
-
-			// If there was a problem, log and move on
-			if err != nil {
-				logger.Warnf("Failed to check blockstore for bitswap entry: %s", entry.Cid)
-				continue
+			if size < sendBlockThreshold {
```
Comment (rvagg) [inline bitswap/provider.go:222]: 👍 nice that we have this now

**Comment by rvagg** [comment]
Looks pretty good to me, nothing obvious standing out as a big problem (_except_ for exacerbating the no-tests issues we have with this codebase). I wouldn't mind seeing that blockstore error log put back in that I've mentioned inline though.

The hard-wired `8` is tempting to make configurable, but I don't really know what alternative numbers would be worth using! the retrieval thread isn't blocking for long so the `8` there is a bit of overkill. However, we're going to be having conversations in the coming weeks about retriever API that may impact this. My personal preference would be making a `Retrieve()` call blocking; in which case, there's opportunity here to make the number of goroutines started for retrieval here be the maximum number of parallel retrievals we're willing to perform (the `AwaitGet` stuff would have to be spawned off in its own goroutine before calling `Retrieve()` and blocking but that would be fine).

File: bitswap/provider.go (hunk 1 above)
Comment (elijaharita) [reply bitswap/provider.go:218]: added

**Comment by elijaharita** [comment]
@rvagg i fixed up the things you pointed out, aside from tests. would it be cool to put tests in a subsequent pr so we can get this merged?

File: bitswap/provider.go [hunk 3]
```diff
@@ -94,264 +121,312 @@ func NewProvider(
...
 				}
-			case topicDontHave:
-				msg.AddDontHave(cid.Cid(topic))
-			case topicBlock:
-				blk, err := provider.blockManager.Get(context.Background(), cid.Cid(topic))
+			} else {
+				log.Infof("Started retrieval for %s", cid)
+			}
+
+			// TODO: if retriever.Request() is changed to be blocking, make
```
Comment (rvagg) [inline bitswap/provider.go:367]: For the record, we're using lassie v0.1.2 which still punts to a goroutine: https://github.com/filecoin-project/lassie/blob/v0.1.2/pkg/retriever/retriever.go#L180

The next update of lassie here will go sync on this call: https://github.com/filecoin-project/lassie/blob/9c4638df01add7d676e53092d1a8d724ba402e59/pkg/retriever/retriever.go#L173

I like the idea of making AwaitBlock cancellable; we'll presumably get the block well before the retrieval has finished (particularly if it's a deep DAG).

File: bitswap/provider.go [hunk 4]
```diff
@@ -26,36 +26,63 @@ import (
...
+
+type ResponseData struct {
+	action ResponseAction
+	reason string
+}
+
+const (
+	sendHave ResponseAction = iota
+	sendDontHave
+	sendBlock
```
Comment (rvagg) [inline bitswap/provider.go:48]: when encountering these values down below I found it a bit confusing what they were; can I suggest we extend their names to explain what they are? `sendHaveAction`, `sendDontHaveAction` or something like that?

**Review by rvagg — APPROVED** [review]
👍 I'm excited to see how this goes; I'm comfortable with you merging it when you're happy with it

**Comment by rvagg** [comment]
CI hopefully fixed by https://github.com/application-research/autoretrieve/pull/166

**Review by hannahhoward — COMMENTED** [review]
Hello I'm really sorry I'm late on the review, cause I might have caught the stuff that broke the provider rewrite.

The short version of what I think is broken is the peertaskqueue includes a freezing feature that's not well documented, that gets triggered whenever a task is removed from the queue due to a cancel. There are basically two ways you can deal with this:

1. Pass the IgnoreFreezing() option when you construct the peer task queues
2. Appropriately thaw the queue to under freezing -- I've put notes where I think this should go, plus references to the original go-bitswap code.

One question overall I'd like to raise: do we really need all these levels of parallelism? And if so, should they all be mediated by the peer task queue rather than simple channels / FIFO queues?

Finally, how can we get some tests in here? We've now broken autoretrieve running live twice by messing with Bitswap Provider code, which isn't totally surprising given that almost all implementations of the bitswap provider are notoriously finicky and complex (a protocol weakness, but one we have to deal with). It seems like we need to have at minimum tests in the code, and possibly a live testing process.

**Comment by elijaharita** [comment]
> Hello I'm really sorry I'm late on the review, cause I might have caught the stuff that broke the provider rewrite.
> 
> The short version of what I think is broken is the peertaskqueue includes a freezing feature that's not well documented, that gets triggered whenever a task is removed from the queue due to a cancel. There are basically two ways you can deal with this:
> 
> 1. Pass the IgnoreFreezing() option when you construct the peer task queues
> 2. Appropriately thaw the queue to under freezing -- I've put notes where I think this should go, plus references to the original go-bitswap code.
> 
> One question overall I'd like to raise: do we really need all these levels of parallelism? And if so, should they all be mediated by the peer task queue rather than simple channels / FIFO queues?
> 
> Finally, how can we get some tests in here? We've now broken autoretrieve running live twice by messing with Bitswap Provider code, which isn't totally surprising given that almost all implementations of the bitswap provider are notoriously finicky and complex (a protocol weakness, but one we have to deal with). It seems like we need to have at minimum tests in the code, and possibly a live testing process.

ok, before any more work on bitswap code, i will add tests to make sure everything is working correctly. 

switching to fifo might be a better solution. i'll explore that once tests are ready.

**Pull request merged**